File: q-reversi-app/lib/domain/services/game_service.dart
import '../entities/game_state.dart';
import '../entities/gate_type.dart';
import '../entities/position.dart';
import '../entities/forbidden_area.dart';
import '../entities/player.dart';
import '../entities/piece_type.dart';
import '../entities/piece.dart';
import '../entities/game_mode.dart';
import 'gate_service.dart';

/// ゲームサービス
class GameService {
  final GateService _gateService = GateService();
  
  /// ゲートを適用（完全な処理）
  GameState applyGateWithFullLogic(
    GameState gameState,
    GateType gate,
    List<Position> targetPositions,
  ) {
    final currentPlayer = gameState.getCurrentPlayer();
    if (currentPlayer == null) return gameState;
    
    // クールタイムチェック（フリーランモードとチャレンジモードではスキップ）
    if (gameState.gameMode != GameMode.freeRun && 
        gameState.gameMode != GameMode.challenge && 
        !currentPlayer.canUseGate(gate)) {
      return gameState; // エラー: クールタイム中
    }
    
    // 禁止領域チェック（1ビットゲートのみ適用、フリーランモードとチャレンジモードではスキップ）
    if (gate.isOneBitGate && 
        gameState.gameMode != GameMode.freeRun && 
        gameState.gameMode != GameMode.challenge) {
      final opponent = gameState.getOpponentPlayer();
      if (opponent != null) {
        final opponentForbiddenAreas = gameState.getForbiddenAreas(opponent.id);
        for (final area in opponentForbiddenAreas) {
          if (_isTargetForbidden(area, gate, targetPositions)) {
            return gameState; // エラー: 禁止領域
          }
        }
      }
    }
    
    // エンタングルメントチェック（仕様: エンタングルされた駒にはゲート操作不可）
    // ただし、1ビットゲートの場合は適用範囲が止まるか、その駒のみスキップされる
    // 2ビットゲートの場合は完全に適用不可
    if (gate.isTwoBitGate) {
      for (final position in targetPositions) {
        final piece = gameState.board.getPiece(position.row, position.col);
        if (piece != null && piece.isEntangled) {
          return gameState; // エラー: エンタングル状態（2ビットゲートは完全に適用不可）
        }
      }
    }
    // 1ビットゲートの場合は、GateService内で適切に処理される
    
    // ゲートを適用
    var newState = _gateService.applyGate(gameState, gate, targetPositions);
    
    // クールタイムを更新（フリーランモードとチャレンジモードではクールタイムを設定しない）
    final updatedPlayer = (gameState.gameMode == GameMode.freeRun || gameState.gameMode == GameMode.challenge)
        ? currentPlayer
        : currentPlayer.useGate(gate);
    final newPlayers = Map<int, Player>.from(newState.players);
    newPlayers[currentPlayer.id] = updatedPlayer;
    
    // フリーランモードまたはチャレンジモードの場合はターンを進めるが、プレイヤーは変更しない
    if (gameState.gameMode == GameMode.freeRun || gameState.gameMode == GameMode.challenge) {
      // クールタイムを減少（自分のクールタイムを減少）
      final decreasedPlayer = updatedPlayer.decreaseCooldowns();
      newPlayers[currentPlayer.id] = decreasedPlayer;
      
      return newState.copyWith(
        players: newPlayers,
        turnCount: newState.turnCount + 1,
        // チャレンジモードとフリーランモードではcurrentPlayerを変更しない
        currentPlayer: gameState.currentPlayer,
      );
    }
    
    // 禁止領域を設定（仕様: 相手の次のターンのみ有効）
    final forbiddenArea = _createForbiddenArea(gate, targetPositions);
    final newForbiddenAreas = Map<int, List<ForbiddenArea>>.from(newState.forbiddenAreas);
    final opponentId = newState.currentPlayer == 1 ? 2 : 1;
    
    // 現在のプレイヤーの禁止領域をクリア（自分のターンが終了したら）
    newForbiddenAreas[currentPlayer.id] = [];
    
    // 相手の禁止領域を設定
    newForbiddenAreas[opponentId] = [forbiddenArea];
    
    // 2ビットゲートが適用された場合、位置を記録（次のターンのプレイヤーに設定）
    final newLastTwoBitGatePositions = Map<int, List<Position>>.from(newState.lastTwoBitGatePositions);
    if (gate.isTwoBitGate && targetPositions.length == 2) {
      // 現在のプレイヤーの記録をクリア
      newLastTwoBitGatePositions[currentPlayer.id] = [];
      // 次のターンのプレイヤー（相手）に記録
      newLastTwoBitGatePositions[opponentId] = List.from(targetPositions);
    } else {
      // 1ビットゲートの場合は、現在のプレイヤーの記録をクリア
      newLastTwoBitGatePositions[currentPlayer.id] = [];
    }
    
    // ターンを進める
    final nextPlayer = newState.currentPlayer == 1 ? 2 : 1;
    final nextPlayerObj = newPlayers[nextPlayer];
    if (nextPlayerObj != null) {
      // 相手のクールタイムを減少
      newPlayers[nextPlayer] = nextPlayerObj.decreaseCooldowns();
    }
    
    // 禁止領域は次のターンのプレイヤー（opponentId）に設定されているので、
    // そのまま使用する（クリアしない）
    
    return newState.copyWith(
      players: newPlayers,
      forbiddenAreas: newForbiddenAreas,
      lastTwoBitGatePositions: newLastTwoBitGatePositions,
      currentPlayer: nextPlayer,
      turnCount: newState.turnCount + 1,
    );
  }
  
  /// ターゲットが禁止領域かどうか
  bool _isTargetForbidden(
    ForbiddenArea area,
    GateType gate,
    List<Position> targetPositions,
  ) {
    if (gate.isOneBitGate) {
      if (area.type == ForbiddenAreaType.row && targetPositions.isNotEmpty) {
        return area.row == targetPositions.first.row;
      }
      if (area.type == ForbiddenAreaType.column && targetPositions.isNotEmpty) {
        return area.column == targetPositions.first.col;
      }
      if (area.type == ForbiddenAreaType.fourPieces) {
        return area.isFourPiecesForbidden(targetPositions);
      }
    }
    return false;
  }
  
  /// 禁止領域を作成
  ForbiddenArea _createForbiddenArea(
    GateType gate,
    List<Position> targetPositions,
  ) {
    if (gate.isOneBitGate) {
      if (targetPositions.length == 8) {
        // 行または列
        if (targetPositions.every((p) => p.row == targetPositions.first.row)) {
          return ForbiddenArea.row(targetPositions.first.row);
        }
        if (targetPositions.every((p) => p.col == targetPositions.first.col)) {
          return ForbiddenArea.column(targetPositions.first.col);
        }
      }
      if (targetPositions.length == 4) {
        // 4マス選択
        return ForbiddenArea.fourPieces(targetPositions);
      }
    }
    // 2ビットゲートには禁止領域なし
    return ForbiddenArea.fourPieces(const []);
  }
  
  /// 初期盤面を生成（ランダム、各タイプ25%）
  GameState createInitialBoard(GameState gameState) {
    final board = gameState.board;
    final totalCells = board.rows * board.cols;
    final piecesPerType = totalCells ~/ 4; // 各タイプ25%
    
    // 各タイプ16個ずつ（8×8の場合）
    final types = [
      PieceType.white,
      PieceType.black,
      PieceType.grayPlus,
      PieceType.grayMinus,
    ];
    final allPositions = <Position>[];
    
    for (int r = 0; r < board.rows; r++) {
      for (int c = 0; c < board.cols; c++) {
        allPositions.add(Position(r, c));
      }
    }
    
    allPositions.shuffle();
    
    // 各タイプのリストを作成
    final typeList = <PieceType>[];
    for (int i = 0; i < types.length; i++) {
      for (int j = 0; j < piecesPerType; j++) {
        typeList.add(types[i]);
      }
    }
    // 残りのマス（割り切れない場合）をランダムに割り当て
    while (typeList.length < totalCells) {
      typeList.add(types[typeList.length % types.length]);
    }
    typeList.shuffle();
    
    var newBoard = board;
    for (int i = 0; i < allPositions.length && i < typeList.length; i++) {
      final position = allPositions[i];
      final type = typeList[i];
      final piece = Piece(
        id: 'piece_${position.row}_${position.col}',
        type: type,
        position: position,
      );
      newBoard = newBoard.setPiece(position.row, position.col, piece);
    }
    
    return gameState.copyWith(board: newBoard);
  }
  
  /// 盤面をすべて白または黒にする
  GameState setAllPiecesToColor(GameState gameState, bool isWhite) {
    final board = gameState.board;
    final pieceType = isWhite ? PieceType.white : PieceType.black;
    
    var newBoard = board;
    for (int r = 0; r < board.rows; r++) {
      for (int c = 0; c < board.cols; c++) {
        final position = Position(r, c);
        final piece = Piece(
          id: 'piece_${r}_$c',
          type: pieceType,
          position: position,
        );
        newBoard = newBoard.setPiece(r, c, piece);
      }
    }
    
    return gameState.copyWith(board: newBoard);
  }
}

File: q-reversi-quantum-minimal/qreversi/game_service.py
"""Full game logic for Q-Reversi.

Ported from q-reversi-app/lib/domain/services/game_service.dart.
Handles:
  - Cooldown enforcement
  - Forbidden-area checks
  - Entanglement guards
  - Turn advancement
  - Legal action enumeration
"""
from __future__ import annotations

from typing import List, NamedTuple, Optional, Tuple

from .types import ForbiddenAreaType, GateType, PlayerColor
from .game_state import (
    ForbiddenArea,
    GameState,
    Player,
)
from .gate_service import GateService, _is_adjacent

_gate_service = GateService()


# ─────────────────────────────────────────────────────────────
# Action representation
# ─────────────────────────────────────────────────────────────

class QRAction(NamedTuple):
    """A legal action: (gate, target_positions)."""
    gate: GateType
    positions: Tuple[Tuple[int, int], ...]  # immutable

    def __repr__(self) -> str:
        return f"QRAction({self.gate.value}, {list(self.positions)})"


# ─────────────────────────────────────────────────────────────
# GameService
# ─────────────────────────────────────────────────────────────

class GameService:
    """Applies a full game step (gate + state bookkeeping)."""

    def apply_action(self, state: GameState, action: QRAction) -> GameState:
        return self.apply_gate(state, action.gate, list(action.positions))

    def apply_gate(
        self,
        state: GameState,
        gate: GateType,
        target_positions: List[Tuple[int, int]],
    ) -> GameState:
        """Apply gate with full VS-mode logic. Returns unchanged state on illegal move."""
        player = state.get_current_player()
        if player is None:
            return state

        # ── Cooldown check ──────────────────────────────────
        if not player.can_use_gate(gate):
            return state

        # ── Forbidden area check (1-bit gates only) ──────────
        if gate.is_one_bit:
            my_forbidden = state.get_forbidden_areas(player.id)
            for area in my_forbidden:
                if _is_target_forbidden(area, gate, target_positions):
                    return state

        # ── Entanglement guard (2-bit gates: hard block) ─────
        if gate.is_two_bit:
            for (r, c) in target_positions:
                p = state.board.get_piece(r, c)
                if p is not None and p.is_entangled:
                    return state

        # ── Apply gate transformation ─────────────────────────
        new_state = _gate_service.apply_gate(state, gate, target_positions)

        # ── Cooldown update ───────────────────────────────────
        updated_player = player.use_gate(gate)
        new_players = dict(new_state.players)
        new_players[player.id] = updated_player

        # ── Forbidden area update ─────────────────────────────
        forbidden_area = _create_forbidden_area(gate, target_positions)
        new_forbidden = {k: list(v) for k, v in new_state.forbidden_areas.items()}
        opponent_id = new_state.opponent_id

        new_forbidden[player.id] = []           # clear current player's restriction
        new_forbidden[opponent_id] = [forbidden_area]  # restrict opponent next turn

        # ── Last 2-bit gate positions ─────────────────────────
        new_two_bit = {k: list(v) for k, v in new_state.last_two_bit_positions.items()}
        if gate.is_two_bit and len(target_positions) == 2:
            new_two_bit[player.id] = []
            new_two_bit[opponent_id] = list(target_positions)
        else:
            new_two_bit[player.id] = []

        # ── Advance turn / opponent cooldown decrease ─────────
        next_player_id = opponent_id
        next_player = new_players.get(next_player_id)
        if next_player is not None:
            new_players[next_player_id] = next_player.decrease_cooldowns()

        return new_state.copy_with(
            players=new_players,
            forbidden_areas=new_forbidden,
            last_two_bit_positions=new_two_bit,
            current_player=next_player_id,
            turn_count=new_state.turn_count + 1,
        )

    # ─────────────────────────────────────────────────────────
    # Legal action enumeration
    # ─────────────────────────────────────────────────────────

    def legal_actions(self, state: GameState) -> List[QRAction]:
        """Return all legal actions for the current player."""
        player = state.get_current_player()
        if player is None or state.is_game_over:
            return []

        actions = []
        board = state.board

        for gate in GateType:
            if not player.can_use_gate(gate):
                continue

            if gate.is_one_bit:
                my_forbidden = state.get_forbidden_areas(player.id)

                # Row targets
                for r in range(board.rows):
                    pos = tuple((r, c) for c in range(board.cols))
                    if not _is_one_bit_target_forbidden(my_forbidden, gate, list(pos)):
                        actions.append(QRAction(gate=gate, positions=pos))

                # Column targets
                for c in range(board.cols):
                    pos = tuple((r, c) for r in range(board.rows))
                    if not _is_one_bit_target_forbidden(my_forbidden, gate, list(pos)):
                        actions.append(QRAction(gate=gate, positions=pos))

                # 4-cell square targets
                for r in range(board.rows - 1):
                    for c in range(board.cols - 1):
                        pos = (
                            (r, c), (r, c + 1),
                            (r + 1, c), (r + 1, c + 1),
                        )
                        if not _is_one_bit_target_forbidden(my_forbidden, gate, list(pos)):
                            actions.append(QRAction(gate=gate, positions=pos))

            else:  # 2-bit gate
                for r in range(board.rows):
                    for c in range(board.cols):
                        for dr in range(-1, 2):
                            for dc in range(-1, 2):
                                if dr == 0 and dc == 0:
                                    continue
                                r2, c2 = r + dr, c + dc
                                if not board.is_valid(r2, c2):
                                    continue
                                p1 = board.get_piece(r, c)
                                p2 = board.get_piece(r2, c2)
                                if p1 is None or p2 is None:
                                    continue
                                if p1.is_entangled or p2.is_entangled:
                                    continue
                                pos = ((r, c), (r2, c2))
                                actions.append(QRAction(gate=gate, positions=pos))

        # Deduplicate (rows/cols produce duplicate pairs for 2-bit gates)
        seen = set()
        unique = []
        for a in actions:
            key = (a.gate, a.positions)
            if key not in seen:
                seen.add(key)
                unique.append(a)

        return unique

    def action_count(self, state: GameState) -> int:
        return len(self.legal_actions(state))


# ─────────────────────────────────────────────────────────────
# Helpers
# ─────────────────────────────────────────────────────────────

def _is_target_forbidden(
    area: ForbiddenArea,
    gate: GateType,
    target_positions: List[Tuple[int, int]],
) -> bool:
    if not gate.is_one_bit:
        return False
    if area.type == ForbiddenAreaType.ROW and target_positions:
        return area.row == target_positions[0][0]
    if area.type == ForbiddenAreaType.COLUMN and target_positions:
        return area.column == target_positions[0][1]
    if area.type == ForbiddenAreaType.FOUR_PIECES:
        return area.is_four_pieces_forbidden(target_positions)
    return False


def _is_one_bit_target_forbidden(
    opp_forbidden: List[ForbiddenArea],
    gate: GateType,
    target_positions: List[Tuple[int, int]],
) -> bool:
    for area in opp_forbidden:
        if _is_target_forbidden(area, gate, target_positions):
            return True
    return False


def _create_forbidden_area(
    gate: GateType,
    target_positions: List[Tuple[int, int]],
) -> ForbiddenArea:
    if gate.is_one_bit:
        n = len(target_positions)
        if n == 8:
            # Row or column?
            rows = {p[0] for p in target_positions}
            cols = {p[1] for p in target_positions}
            if len(rows) == 1:
                return ForbiddenArea.for_row(next(iter(rows)))
            if len(cols) == 1:
                return ForbiddenArea.for_column(next(iter(cols)))
        if n == 4:
            return ForbiddenArea.for_four_pieces(target_positions)
    # 2-bit gate or fallback: empty forbidden area (no restriction)
    return ForbiddenArea.for_four_pieces([])
